flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java

```diff
@@ -81,4 +86,13 @@ public boolean tryYield() {
 			return false;
 		}
 	}
+
+	private Mail compose(@Nonnull Runnable command, int priority, String descriptionFormat, Object[] descriptionArgs) {
+		return new Mail(command, priority, descriptionFormat, descriptionArgs) {
```

Comment:
> Also, as it is now, Mail class is not used anywhere in the production code, which makes it a bit strange. If we want to always use actionExecutor, then it should be added to the Mail.

Added it to `Mail`, thanks.


> I would try to avoid wrapping Mail into a no-op actionExecutor if it's possible

Why? 
Until we move the `getCheckpointLock()` method to `SourceStreamTask` it's always **not** no-op. After that, we can consider passing null/none (though I doubt if it's needed; I did a quick benchmark and didn't see any significant overhead).

> Maybe shortcutting this wrapping if actionExecutor == IMMEDIATE?

This could easily break.